### PR TITLE
Dashboard to work with Server APIs 

### PR DESCRIPTION
**Pull request by npalaska**
This is a followup improvement PR on top of https://github.com/distributed-system-analysis/pbench-dashboard/pull/75

> Replaced all Elasticsearch calls to use pbench server APIs
Server address {pbench_server) needs to be specified in datastoreConfig.js to enable API calls
References for Elasticsearch and GraphQL from datastoreConfig.js have been moved to the server(pbench-server.cfg)

### CODE REVIEW

**Comment by npalaska** [comment]
How do we update the `datastoreConfig` file for the travis CI runs?

**Comment by aquibbaig** [comment]
> How do we update the `datastoreConfig` file for the travis CI runs?

Hi @npalaska. I dont think you can update or you need to do that. The endpoints contained in datastoreConfig are actually replaced by mock files in the test suite. Also, the tests are failing, let's debug them together

**Comment by npalaska** [comment]
@aquibbaig are you talking about the `e2etests` file where we export the `datastoreConfig` values to `mock/datastoreConfig.js`,? but in that case we would have to expose our internal `pbench_server` endpoint isnt it

**Comment by aquibbaig** [comment]
> @aquibbaig are you talking about the `e2etests` file where we export the `datastoreConfig` values to `mock/datastoreConfig.js`,? but in that case we would have to expose our internal `pbench_server` endpoint isnt it

Yes @npalaska. The datastoreConfig remains private as your own local configuration. We don't commit it anywhere, in e2e tests the values need to be fetched from the mock file rather than any endpoints specified in the `datastoreConfig.js`. Considering the fact that now the urls have been modified after migrating to pbench server, you might wanna check if we are still able to intercept requests in each page and replace the data contained in those urls with those present in our mock file.

File: src/services/search.js [hunk 1]
```diff
@@ -4,56 +4,61 @@ import request from '../utils/request';
 export async function queryIndexMapping(params) {
   const { datastoreConfig, indices } = params;
 
-  const endpoint = `${datastoreConfig.elasticsearch}/${datastoreConfig.prefix}${
-    datastoreConfig.run_index
-  }${indices[0]}/_mappings`;
+  const url = `${datastoreConfig.prefix}${datastoreConfig.run_index}${indices[0]}/_mappings`;
```
Comment (dbutenhof) [inline src/services/search.js]: This isn't a `url`. It's not really an "index" either since you're pulling the template mappings for one index.

I'm not quite sure _what_ to call this, actually; although in terms of the interim pass-through Elasticsearch API, we call it "indices".

**Review by dbutenhof — COMMENTED** [review]
We need to fix up the Ansible dashboard deploy/install mechanism to correctly generate the dashboard config file from an updated `config.json.j2` template.

I'm also not that thrilled about separating this "essential cleanup" from the original PR, but that's a technicality.

File: src/services/search.js (hunk 1 above)
Comment (portante) [reply src/services/search.js]: Perhaps we should separate this out a bit:

 * The code forms a full index name: `${datastoreConfig.prefix}${datastoreConfig.run_index}${indices[0]}`
 * The `url` is really trying to specify the `_mapping` API for a given index

**Review by portante — CHANGES_REQUESTED** [review]

File: src/e2e/controllers.e2e.js [hunk 2]
```diff
@@ -37,7 +37,7 @@ beforeAll(async () => {
         headers: { 'Access-Control-Allow-Origin': '*' },
         body: JSON.stringify(generateMockControllerAggregation),
       });
-    } else if (request.method() === 'GET' && request.url().includes('aliases')) {
+    } else if (request.method() === 'POST' && request.postData().includes('aliases')) {
```
Comment (dbutenhof) [inline src/e2e/controllers.e2e.js]: I'd prefer the JSON payload check to be a bit more specific than just having "aliases" anywhere. Though most likely the test environment won't ever see it anywhere but the expected "data"... and, OK, the adjacent tests aren't any more specific. :-(

The repetition of this pattern so many times seems awkward, too; what's really happening is trying to identify a query from the HTTP method and a "unique" URI or some snippet of JSON payload. This could probably be refactored and cleaned up (and made more robust); but despite griping about it, I'm not going to push you to attempt that now.

File: src/services/search.js [hunk 3]
```diff
@@ -4,11 +4,13 @@ import request from '../utils/request';
 export async function queryIndexMapping(params) {
   const { datastoreConfig, indices } = params;
 
-  const url = `${datastoreConfig.prefix}${datastoreConfig.run_index}${indices[0]}/_mappings`;
+  const mappingIndix = `${datastoreConfig.prefix}${datastoreConfig.run_index}${
```
Comment (dbutenhof) [inline src/services/search.js]: Oops; `mappingIndex`??

File: src/e2e/controllers.e2e.js (hunk 2 above)
Comment (npalaska) [reply src/e2e/controllers.e2e.js]: The JSON payload is just a small mock payload that we return for the purpose of e2e tests and I am not sure but I think that might have been the original reason behind this checks, I can definitely clean this up a little but this is again going to be modified when we have query specific REST APIs

File: src/services/search.js (hunk 3 above)
Comment (npalaska) [reply src/services/search.js]: yeah need to think something better

Comment (aquibbaig) [reply src/services/search.js]: Simply, `mappings` should be fine

**Review by aquibbaig — APPROVED** [review]
Looks good, @npalaska! Great work! :tada:

File: src/services/search.js (hunk 3 above)
Comment (dbutenhof) [reply src/services/search.js]: I was really just commenting on the typo. I wouldn't object to a more clear name, if anyone can think of one.

**Review by dbutenhof — APPROVED** [review]

File: src/e2e/controllers.e2e.js (hunk 2 above)
Comment (portante) [reply src/e2e/controllers.e2e.js]: What is the reason for the change from `GET` to `POST`?

**Review by portante — APPROVED** [review]
Just wondering about the `GET` to `POST` reason ...

**Comment by npalaska** [comment]
@portante Its in accordance with the changes we have in our Server API, if you look at the `/elasticsearch` endpoint in server API there is no `GET` method only `POST` request. We decide whether to do the get request based on the condition of payload in the data, if payload is present in the data we perform the post otherwise if payload is missing then we just perform the get request and return the response we got from the elasticsearch 
To which I think should have had a proper `GET` method in our API but we are anyway moving away from the API calls to direct `elasticsearch` endpoint to support more targeted queries

File: src/services/dashboard.js [hunk 4]
```diff
@@ -38,41 +41,46 @@ export async function queryControllers(params) {
...
-      },
       data: {
-        filter: {
-          range: {
-            '@timestamp': {
-              gte: selectedDateRange.start,
-              lte: selectedDateRange.end,
-            },
-          },
+        indices,
```
Comment (dbutenhof) [inline src/services/dashboard.js]: This is more a comment on the original Pbench server API design, but it didn't really start bugging me until I saw it appear repeatedly in the dashboard changes.

This is not well-formed JSON, because we're just throwing in an unlabeled string. This ought to be another key, like "uri" ... although `uri` is misleading since it's really only the resource part of the uri. So I suppose `"resource"` or `"target_resource"` would be better.

I'm not inclined to push this issue too far, since these changes are short lived, and when we have a real API we'll be cutting off the pass-through APIs anyway. Still, it's going to continue to bug me every time I see it.

**Review by dbutenhof — APPROVED** [review]

File: src/services/dashboard.js (hunk 4 above)
Comment (npalaska) [reply src/services/dashboard.js]: I was trying to make eslint happy with https://eslint.org/docs/rules/object-shorthand. Apparently in ES6 `var foo = {x, y, z};` = `var foo = {x:x, y:y, z:z};`

Comment (dbutenhof) [reply src/services/dashboard.js]: Ah; and I knew that, too, but I clearly hadn't completely twisted my brain from Python to Javascript! Sorry 'bout that.

On the other hand, a JSON key of "indices" isn't really accurate, either, but that's not a big deal in this context.

**Review by gurbirkalsi — APPROVED** [review]

**Review by anishaswain — APPROVED** [review]

**Comment by dbutenhof** [comment]
This really needs to be rebased and brought up to date!

File: src/services/dashboard.js [hunk 5]
```diff
@@ -7,11 +7,15 @@ import { getAllMonthsWithinRange } from '../utils/moment_constants';
 const { endpoints } = window;
 
 function scrollUntilEmpty(data) {
-  const endpoint = `${endpoints.elasticsearch}/_search/scroll?scroll=1m`;
+  const url = `/_search/scroll?scroll=1m`;
```
Comment (dbutenhof) [inline src/services/dashboard.js]: I don't think `url` is used, and it's redundant with `indices` below.

File: src/services/search.js [hunk 6]
```diff
@@ -6,55 +6,61 @@ const { endpoints } = window;
...
 
-  const endpoint = `${endpoints.elasticsearch}/${endpoints.prefix}${endpoints.run_index}${
-    indices[0]
-  }/_mappings`;
-  return request.get(endpoint);
+  const mappings = `${endpoints.prefix}${endpoints.run_index}${indices[0]}/_mappings`;
+
+  const endpoint = `${endpoints.pbench_server}/elasticsearch`;
+
+  return request.post(endpoint, { data: { indices: mappings } });
```
Comment (dbutenhof) [inline src/services/search.js:13]: Why not just call it `indices` instead of `mappings`? (I don't think the key "indices" is really any more accurate anywhere else than here, and it's just a name.)

But, no big deal, and no need to change it.

File: src/services/dashboard.js [hunk 7]
```diff
@@ -258,50 +287,55 @@ export async function queryIterationSamples(params) {
 export async function queryTimeseriesData(payload) {
   const { selectedDateRange, selectedIterations } = payload;
 
-  const endpoint = `${endpoints.elasticsearch}/${getAllMonthsWithinRange(
+  const indices = `${parseMonths(
```
Comment (dbutenhof) [inline src/services/dashboard.js]: Why the change from `getAllMonthsWithinRange` to `parseMonths`?

Comment (dbutenhof) [reply src/services/dashboard.js]: > Why the change from `getAllMonthsWithinRange` to `parseMonths`?

Ah; and that's your unit test failure, because there _is_ no `parseMonths`.

Comment (npalaska) [reply src/services/dashboard.js]: This is interesting I dont seem to recall how this change came while rebasing

**Review by portante — COMMENTED** [review]
General direction looks good.  I'll review again once @dbutenhof's comments have been addressed.

**Comment by aquibbaig** [comment]
Hey @npalaska! Can you rebase this branch on top of `master` which will probably fix those tests for you

File: src/services/dashboard.js (hunk 5 above)
Comment (portante) [reply src/services/dashboard.js]: @npalaska ?

File: src/services/dashboard.js (hunk 7 above)
Comment (portante) [reply src/services/dashboard.js]: @npalaska ?

**Review by portante — CHANGES_REQUESTED** [review]
Needs a rebase and outstanding comments addressed and resolved.

**Review by riya-17 — APPROVED** [review]
Looks Good

**Comment by dbutenhof** [comment]
Can we pick this up again? It should be close to ready with a rebase and minor adjustments; it would get the dashboard away from direct Elasticsearch access. (The GraphQL direct access will go away in any case as we migrate to Pbench server authentication and metadata, but this seems like a good interim step.)

File: src/services/dashboard.js [hunk 8]
```diff
@@ -90,112 +98,127 @@ export async function queryResults(params) {
...
+                },
+                aggs: {
+                  type: {
+                    terms: {
+                      field: 'sample.measurement_type',
+                    },
+                    aggs: {
+                      title: {
+                        terms: {
+                          field: 'sample.measurement_title',
```
Comment (dbutenhof) [inline src/services/dashboard.js]: You lost the `.raw` here; `sample.measurement_title` isn't "aggregatable" (`"text"`) but `sample.measurement_title.raw` is (`"keyword"`)

**Review by dbutenhof — CHANGES_REQUESTED** [review]
Mostly good, but one critical transcription error.

File: src/services/dashboard.js (hunk 8 above)
Comment (npalaska) [reply src/services/dashboard.js]: Ah good catch

**Review by dbutenhof — APPROVED** [review]
Reminder... this would have been easier to re-review if you'd pushed a separate commit to fix the problem, which would allow reviewers to be more confident they were seeing all changes without re-reviewing _everything_! ;-)

**Comment by npalaska** [comment]
> Reminder... this would have been easier to re-review if you'd pushed a separate commit to fix the problem, which would allow reviewers to be more confident they were seeing all changes without re-reviewing _everything_! ;-)

Yes; it was just 1 line change so I decided to squash it in the previous commit. I also didnt want to get approval and then discard all the approvals because of the rebase, but interestingly it didnt happen here with my previous rebase, huh !!

File: src/services/dashboard.js [hunk 9]
```diff
@@ -7,11 +7,14 @@ import { getAllMonthsWithinRange } from '../utils/moment_constants';
 const { endpoints } = window;
 
 function scrollUntilEmpty(data) {
-  const endpoint = `${endpoints.elasticsearch}/_search/scroll?scroll=1m`;
+  const endpoint = `${endpoints.pbench_server}/elasticsearch`;
   const allData = data;
+  const indices = `_search/scroll?scroll=1m&scroll_id=${allData._scroll_id}`;
```
Comment (portante) [inline src/services/dashboard.js]: Usually, the first `scroll` API call established a context, returning a `scroll_id`, which is used on subsequent calls.  Is that still happening?

Comment (npalaska) [reply src/services/dashboard.js]: Not sure what you meant by that, this is just a structural change separating an es endpoint from indices without any logical changes. In the original code we are querying like this
```
const endpoint = `${endpoints.elasticsearch}/_search/scroll?scroll=1m`;
const scroll = request.post(`${endpoint}&scroll_id=${allData._scroll_id}`);
```
However with server api we need a `data` field with a post payload so we do the following specifying the indices separately 
```
endpoint = `${endpoints.pbench_server}/elasticsearch`;
const indices = `_search/scroll?scroll=1m&scroll_id=${allData._scroll_id}`;
const scroll = request.post(endpoint, {
      data: { indices },
    });

Comment (portante) [reply src/services/dashboard.js]: Isn't there an `if` on line 13 where the `scroll_id` is only added if it exists in the returned data?

The `scroll_id` seems to only be added at line 19 below.

**Review by portante — CHANGES_REQUESTED** [review]
Seems like we need to correct the use of the `scroll_id` still, but looks good otherwise.

**Comment by dbutenhof** [comment]
> Seems like we need to correct the use of the `scroll_id` still, but looks good otherwise.

Need to figure out whether this is actually the case and if so, get in a fix. I'd like to get this merged finally, if we can!

File: src/services/dashboard.js (hunk 9 above)
Comment (dbutenhof) [reply src/services/dashboard.js]: I think that Javascript will just return `undefined` at line 12 if the `_scroll_id` property is missing, and in that case the `if` at line 14 won't trigger, so the `&scroll_id=undefined` won't be used.

But perhaps Gurbir or Aquib can confirm the Javascript behavior more confidently...

Comment (dbutenhof) [reply src/services/dashboard.js]: Actually, I just pulled up Chrome's debug console and tried it; the undefined JSON field does simply resolve as "undefined":

```
x = {a: "test", b: "foo"}
y = `?scroll_id=${x.bar}`

"?scroll_id=undefined"
```

So I'm pretty sure this is fine.

Comment (portante) [reply src/services/dashboard.js]: What does the Elasticsearch V7 scroll API do when it receives `undefined` as the `scroll_id`?

Aside from the fact that specifying the `scroll_id` in the parameters of the URL is deprecated as of 7.0, has anybody tested this change to make sure it behaves as expected?

See https://www.elastic.co/guide/en/elasticsearch/reference/current/scroll-api.html.

Comment (dbutenhof) [reply src/services/dashboard.js]: But at line 14, `allData._scroll_id` will be False and we won't POST the request, so that doesn't matter.

On the other hand, sure, the code would be a tiny bit more efficient if we assembled the potential URI within the conditional.

Comment (npalaska) [reply src/services/dashboard.js]: Okay I fixed it by creating indices inside if block so `scroll_id` should never be undefined in the `indices`

Comment (portante) [reply src/services/dashboard.js]: Okay, it appears that `scrollUntilEmpty` is only ever called after an initial `_search/scroll` API is made, so this appears to be correct.

Though, it would be nice if somebody can confirm this actually still works.

**Review by portante — APPROVED** [review]
Let's hope the scroll API still works ...

File: src/services/dashboard.js (hunk 9 above)
Comment (dbutenhof) [reply src/services/dashboard.js]: I just did some experimenting with direct Elasticsearch queries, and id `15ed6b30e0f92df4fffdba7bae72c41d` (`fio_rw_2021.03.03T02.04.39`) on the staging server is returning a `_scroll_id` for `queryIterationSamples`, so the code ought to be triggering if Nikhil runs his dashboard code and drills down through that dataset.

**Comment by gurbirkalsi** [comment]
@npalaska and I discovered an issue where the `ignore_unavailable:true` parameter for each request to Elasticsearch was not encoded correctly within the request URL to the Pbench API server. Nikhil is going to open a pull request on the server to address this encoding issue. We were able to visualize data for various workflows where data is available on staging and can merge this PR!
